File: 300-longest-increasing-subsequence/longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        vector<int> dp(n, 1);
       
        for (int i = 1; i < n; i++) {
            for (int j = i; j >= 0; j--) {
                if (nums[i] > nums[j]) dp[i] = max(dp[i], dp[j]+1);
            }
        }
        for (auto i : dp) cout << i << " ";
        return *max_element(dp.begin(), dp.end());
    }
};

File: 300-longest-increasing-subsequence/longest-increasing-subsequence.py
class Solution(object):
    def lengthOfLIS(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if len(nums) == 0:
            return 0
        dp = [1] * len(nums)
        # dp = [1 1 1 1 ...]

        # range(0, len(nums), 1) # start index, end index, step
        # range(len(nums)-1, -1, -1)
        for i in range(len(nums)-1, -1, -1):
            for j in range(i+1, len(nums)):
                if nums[i] < nums[j]:
                    dp[i] = max(dp[i], dp[j]+1)

        return max(dp)

